### ISSUE TITLE
Make CompiledContainer idempotent?

### ISSUE THREAD

**Issue by holtkamp** [PHP-DI/PHP-DI#604]
Heroku [offers free "dyno's" / containers, which go to "sleep" after idling 30 minutes](https://devcenter.heroku.com/articles/free-dyno-hours), in the logs this looks like this:

```
2018-06-08T13:36:14.928015+00:00 heroku[web.1]: Idling
2018-06-08T13:36:14.928583+00:00 heroku[web.1]: State changed from up to down
2018-06-08T13:36:15.659288+00:00 heroku[web.1]: Stopping all processes with SIGTERM
2018-06-08T13:36:15.772020+00:00 heroku[web.1]: Process exited with status 143
2018-06-08T13:36:15.671374+00:00 app[web.1]: Going down, terminating child processes...
2018-06-11T06:38:43.485395+00:00 heroku[web.1]: Unidling
2018-06-11T06:38:43.485781+00:00 heroku[web.1]: State changed from down to starting
2018-06-11T06:38:52.247525+00:00 heroku[web.1]: Starting process with command `vendor/bin/heroku-php-apache2 public/`
2018-06-11T06:38:56.892979+00:00 app[web.1]: DOCUMENT_ROOT changed to 'public/'
2018-06-11T06:38:56.910706+00:00 app[web.1]: Using WEB_CONCURRENCY=2 processes.
2018-06-11T06:38:56.918291+00:00 app[web.1]: Starting php-fpm...
2018-06-11T06:38:58.919763+00:00 app[web.1]: Starting httpd...
2018-06-11T06:38:59.300606+00:00 heroku[web.1]: State changed from starting to up
```

After waking up, a problem seems to occurs with the combination of:

- PHP-DI CompiledContainer
- [OPCache file cache](http://php.net/manual/en/book.opcache.php)
  - Note that [OPCache file cache](http://php.net/manual/en/opcache.configuration.php#ini.opcache.file-cache) is enabled: OPCache entries are also saved to disk, so upon server start they are available immediately. Not sure what parts of this Heroku "keeps" between idling / unindling...

```
Array
(
    [type] => 1
    [message] => Uncaught Error: Call to undefined method CompiledContainer::get5b1e190a4fff9261052747() in /app/vendor/php-di/php-di/src/CompiledContainer.php:53
Stack trace:
#0 /app/data/cache/CompiledContainer.php(39133): DI\CompiledContainer->get('settings.amazon...')
#1 /app/data/cache/CompiledContainer.php(39143): CompiledContainer->get5b1e190b4c240336720486()
#2 /app/vendor/php-di/php-di/src/CompiledContainer.php(53): CompiledContainer->get5b1e190b4c17c596082894()
#3 /app/data/cache/CompiledContainer.php(39098): DI\CompiledContainer->get('Aws\\Credentials...')
#4 [internal function]: CompiledContainer::{closure}(Object(CompiledContainer))
#5 /app/vendor/php-di/invoker/src/Invoker.php(82): call_user_func_array(Object(Closure), Array)
#6 /app/vendor/php-di/php-di/src/CompiledContainer.php(116): Invoker\Invoker->call(Object(Closure), Array)
#7 /app/data/cache/CompiledContainer.php(39099): DI\CompiledContainer->resolveFactory(Object(Closure), 'awsClientOption...')
#8 /app/vendor/php-di/php-di/src/CompiledContainer.php(53):
    [file] => /app/vendor/php-di/php-di/src/CompiledContainer.php
    [line] => 53
)
```

I am not sure, but I got the feeling that this has to do with the CompiledContainer that is regenerated when the dyno comes back to live, but the OPCode cache is not cleared and still refers to the "old" version of the CompiledContainer. 

This can probably be circumvented when the CompiledContainer is made [idempotent](https://en.wikipedia.org/wiki/Idempotence): "a specific PHP-DI configuration X always results in CompiledContainer Y". Currently some random strings are used in the CompiledContainer which causes each regeneration/compilation of the container to have a different result, while no change in the configuration is made. For example:

- [when determining the methodName](https://github.com/PHP-DI/PHP-DI/blob/cdcf21d2a8a60605e81ec269342d48b544d0dfc7/src/Compiler/Compiler.php#L152)
- [when determining the subEntryName](https://github.com/PHP-DI/PHP-DI/blob/cdcf21d2a8a60605e81ec269342d48b544d0dfc7/src/Compiler/Compiler.php#L264)

Would it be an idea to use a hash to determine these strings so that the result would be the same over and over again? Maybe the content of the Definition should also be taken into account so that "changes" to it result in different names...

**Comment by holtkamp** [comment]
>And just to be sure I understand: on Heroku you are letting the container compile itself on the first HTTP hit? You do not generate the compiled container while deploying? Because if you generate the compiled container beforehand that may solve that problem. 

Indeed, there is a catch though 😉 

We do some heavy lifting when deploying like creating the proxies, but the CompiledContainer is placed in a (cache) folder that is eventually removed. The reason for this is that we use a DTAP-approach. After developing locally, we deploy to a `test` environment. Then [Heroku Pipelines](https://devcenter.heroku.com/articles/pipelines) allows us to promote this deployment to another environment like `acceptation` / `staging` and eventually to `production`. 

Each of these environments have different settings (database URL, credentials, etc). So we can not re-use the CompiledContainer... This would cause settings for the `test` environment, eventually end up in `acceptation` and `production` as well. 

In summary, our PHP-DI configuration files are "environment"-specific and therefore: so is the CompiledContainer.

**Comment by mnapoli** [comment]
Ok I see, that makes sense!

**Comment by holtkamp** [comment]
**UPDATE: caches are empty when dyno wakes up**
It seems the OPCache is not involved with the reported problem. After an idling dyno has started up, the OPCache is empty. For what its worth, also the APCu cache is empty. 


A possible cause of the problem might be a "racing condition"? The involved PHP-DI configuration is quite big. Therefore the CompiledContainer is also quite big and it takes some time to compile: 
- 2242036 bytes
- 51078 lines of code

In case multiple requests are processed at "the same time" when the container is not compiled yet, this might cause the reported behaviour, right? This might also explain why the reported behaviour occurs "sometimes" and not every time a dyno wakes up. 

Using an idempotent container should be able prevent any problems like this as well 😄 .

**Comment by holtkamp** [comment]
Thanks @sridesmet yeah, that works nice. 

The "problem" is with [Heroku Pipelines](https://devcenter.heroku.com/articles/pipelines) which promotes:
- `test` to `acceptation`
- `acceptation` to `production

This all uses the same code base. The "problem" can be circumvented by ONLY using settings as defined in environment variables. Unfortunately, some of these settings are still hard coded in the described situation. Which of course is bad practice and on the to-do list to get rid of...

**Comment by mnapoli** [comment]
Since I've been playing a lot with lambdas lately (and Docker containers can be a bit similar) environment variables make a lot of sense for that :)

However sometimes in development or acceptance environments there are whole config/services that are replaced (e.g. mock emails, message queues, etc.) so it still makes sense to me to have different config files. With an `APP_ENV` environment variable you can configure which config files to load. And we are back to your problem:

> a specific PHP-DI configuration X always results in CompiledContainer Y

Your suggestion of idempotency still makes a lot of sense IMO ^^

**Comment by holtkamp** [comment]
>However sometimes in development or acceptance environments there are whole config/services that are replaced (e.g. mock emails, message queues, etc.) 

Yeah indeed, got some pretty big config files here as well. Not sure whether environment variables are the best way to replace these guys 🤓 .
